### ISSUE TITLE
*** Snapshots and Pre-Releases for FlexibleAdapter v5.0.0 ***

### ISSUE THREAD

**Issue by davideas** [davideas/FlexibleAdapter#39]
This is an open discussion to keep track of all the work in progress of the next coming pre-releases in order to arrive to a stable and final version.

In this major release (5.0.0) there will be lot of new features, while trying to simplify the configuration to speed up the development. A full list of the features and improvements, with their status, can be found here [Milestone 5.0.0](https://github.com/davideas/FlexibleAdapter/milestones/5.0.0). However we can identify them in:
- Expandable item with selection coherence.
- Predefined ViewHolders.
- Adapter Animations.
- Drag&Drop and Swipe actions.
- New concept of Item.
- Simplify the constructor.
- New FastScroller to include in the library.
- Rewrite filtering and restore methods.
- Headers/Sections
- etc..

In order to do them, I am completely changing and refactoring the code, names and how the Adapter will be extended to override its methods.

### Gradle configuration
```
allprojects {
    repositories {
        // For [pre-]releases
        jcenter()
        // For SNAPSHOT
        maven { url "https://oss.sonatype.org/content/repositories/snapshots/" }
    }
}
```
```
// Don't cache SNAPSHOT (changing) dependencies.
configurations.all {
    resolutionStrategy.cacheChangingModulesFor 0, 'seconds'
}
```
```
dependencies {
    // JCenter for PreReleases
    compile 'eu.davidea:flexible-adapter:5.0.0-rc3'
    compile 'eu.davidea:flexible-adapter-livedata:1.0.0-b1'
    compile 'eu.davidea:flexible-adapter-databinding:1.0.0-b2'

    // Maven for SNAPSHOT
    compile ('eu.davidea:flexible-adapter:5.0.0-SNAPSHOT') {
        changing = true;
    }
}
```
As soon as a new SNAPSHOT is ready and usable I will communicate it here.

### Suggestions
- Feel free to **Fork** but better to advise me in advance what are you going to modify in order to coordinates the commits and the work.
- Also, feel free to test and report back here or by opening a new Issue.

### Naming convention adopted
| Version | Reason for publishing | Description |
| --- | --- | :-- |
| `5.0.0-SNAPSHOT` | continuous development | A library with many changes that might be unstable, nice to have to early benefit of the latest bug fix, improvements and new features as soon as they are implemented.<br>It will replace the existent library (with that name) in the maven snapshot repository. So you don't have to manually change the version in the dependencies, but expect to upgrade your code in order to continue to compile your project.<br/><br/>_Gradle_ has internal caching system with default time of 24h. Passed this time it will fetch the new _Snapshot_ file from repository, if a new exists. |
| `5.0.0-bX` | beta pre-release | Usable beta library with some new features to test. |
| `5.0.0-rcX` | release candidates | Nearly completed library with good stability and improved features |
| `5.0.0` | release | The stable library with all new features confirmed and tested. |

**Comment by davideas** [comment]
While importing the library v5.0.0-b2 in another project, I got some difficulties implementation. Please be patient since I am working hard to simplify them for next beta: I'm changing the signature of the classes and reviewing the Item interfaces.

**Comment by davideas** [comment]
New pre-release [v5.0.0-b3](https://github.com/davideas/FlexibleAdapter/releases/tag/5.0.0-b3) with new functionalities, many fixes and improvements.
Now it is really usable. Suggestions are welcome.
```
dependencies {
	compile 'eu.davidea:flexible-adapter:5.0.0-b3'
}
```

For beta4:
- I will try to add Maven Snapshots.
- I will focus on _header/sections_ methods to work in combination with add/delete/move item that is attached to.
- Probably, I will add sticky functionality by importing and improving this library [timehop/sticky-headers-recyclerview](https://github.com/timehop/sticky-headers-recyclerview) into mine.
- I will evaluate to create more interfaces like _IFilterable, IAnimable_ for Items.
- I will evaluate to add a lower level of Adapter class to regroup and harmonize some setters for configuration.
- I would like to review the Restore, because it doesn't work well in all situations.

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** with:

**Fixes**
- Fixed crash problem with SelectAll and improved method logic.
- Fixed last expandable item in _expandAll()_.
- Restoring all items, now it's much faster in case of restoreSelection is enabled.

**Headers**
- Headers are shown/hidden during expand/collapse if they belongs to children.
- Headers work in combination with restore items, in all situations (... the ones I have foreseen).
- Headers can also be deleted and restored in any situations: they are relinked all the time with the item below it.
- Headers can be dragged, swapped and receive the new swapped item.
- Support to swap headers only.

**Others additions and improvements**
- FlexibleAdapter is not anymore abstract.
- Added support for AndroidStudio 2.0 InstantRun: updated Gradle plugin.
- Added support for _isSelectable()_ it should now be usable.
- Added methods _addSubItems()_ and _addAllSubItemsFrom()_.
- Transformed all "add" methods in a chain of calls. Modified the signature of _getCurrentChildren()_ to accept an IExpandable object.
- Added minimal support for _isEnabled()_.
- Completely changed the logic how a removed item saves restoration info (now a removed item has an internal reference that should never fail, _adjustRemoved()_ became obsolete and has been removed).
- Included in the library Utils class the method _setHighlightText()_ Useful to highlight text coming from a search action.
- Made a better use of Payload when _notifyItemChanged_ is triggered.

**Comment by davideas** [comment]
**Note:** Currently Headers are _ISectionable_ objects containing the attached item which is updated if an event on this item occurs, the headers are scanned to identify if one of them has that item and so it is updated.
I'm thinking/working to invert how a header is recognized and assigned, in such way: any item can be an ISectionable object containing the reference to another item that WILL BE the Header.
This could simplify the logic and reduce bugs and maintenance, I need to analyse this point

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. Related Apk is avalable under the [latest release](https://github.com/davideas/FlexibleAdapter/releases).

**Refactor**
_IFlexibleItem_ to _IFlexible_. Sorry for this, but it's now more coherent with the others interfaces names.

**Improvements**
Cast optimization: removed unnecessary 8 casts by changing some parameters of methods from generic object T to IFlexible and IExpandable types, this should be transparent.

**Headers**
- Finally, I have evaluated better the _Headers/Sections_ items and decided to invert how a header is created and recognized by the Adapter. Benefits of this change:
>It reduces maintenance and potential future bugs,
>It should have higher performance 
>Less memory usage (removed the internal list of headers).

- _IHeader_ interface has been added to identify the fact the item is a Header item.
- _ISectionable_ interface now has to be assigned to the items(!!) in order for them to hold the reference to the _IHeader_ item.
- Dragging items, it automatically re-links the header in almost all situations (the logic can be improved).
- Deleting all items under a header, when the headers are visible and hidden, it makes the **header orphan**, an option has been added to delete the orphan headers as well. Currently the orphan headers, ones hidden, cannot be shown anymore (they are lost, food for the gc), you should re-link it manually (I might create a list of orphan headers and create some methods around that list, I need to analyse this point).

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. Related Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases).

**Filter**
Added _IFilterable_ item interface. New clean way to filter items, Items now can implement filter() method in order to be collected for the filtered list.

**Headers**
Added support for Orphan Headers.

**Fixes**
Fixed hidden header silent un-linkage when deleting an item.

**Comment by davideas** [comment]
New pre-release [v5.0.0-b4](https://github.com/davideas/FlexibleAdapter/releases/tag/5.0.0-b4) with new functionalities (Sections and Sticky headers included) and lot of improvements. (You can download the apk from the release page).

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published.

**Headers**
- Now headers are checked to be shown when adding items.
- Added _getItemCountOfTypesUntil()_ useful to retrieve the correct position where to insert the new items.

**Comment by farfromrefug** [comment]
Wanted to say how amazing that lib is!
Just discovered it and already digging it a lot.
I forked it so will try to help as much as i can.
Thanks

**Comment by davideas** [comment]
@farfromrefug Thanks a lot! I'm working hard.

Feel free to Fork but be sure to modify the **dev** branch, Also, advise me in advance what are you going to modify in order to coordinates the commits and the work.

**Comment by BoldijarPaul** [comment]
Hello. I think that it would be better if the examples are completely separated.
Currently i'm trying to understand how the sticky headers work in the beta version, and it's pretty hard.

Awesome lib btw.

**Comment by davideas** [comment]
Yes @BoldijarPaul, another Activity showing how headers/sections work has to be done, and not only that, however, I have very few time in this period, maybe in April or later.

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. Related Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases).

**Refactor**
`setHeader(header)` in `ISectionable` interface is changed to **void** method.

**Headers**
- Now headers are clickable when sticky! StickyHeaders can now benefit of the existing click listeners in FlexibleViewHolder.
- `StickyHeaderDecoration` is now deprecated, as consequence `enableStickyHeaders(maxCachedHeaders)` is also deprecated: use `enableStickyHeaders()`.
- Pay attention at the method `getStickySectionHeadersHolder()` (comment is exhaustive).

**Improvements**
Multiple code improvements with **SonarQube** tool, thanks to @georgekankava

**Fixes**
Fixed NPE in `isEnabled(position)` when position is unknown.

**Comment by maxoreli** [comment]
Good job davideas.

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published.

**Refactor**
Refactored constructor of `AbstractSectionableItem` that now accepts the `IHeader` item.

**Headers**
- The Adapter now supports an `IHeader` item combined with `IExpandable` item.
- Added `AbstractExpandableHeaderItem`, an Expandable section item with Header ready to be used.
- Added a new Wiki page about how to implement sections with sticky headers and expandable sections. Direct link: [Headers and Sections](https://github.com/davideas/FlexibleAdapter/wiki/5.x-%7C-Headers-and-Sections).

**Fixes**
- Fixed header linkage when deleting/restore items with header. _It's important now that all sectionable items must initialize the own header at startup_.
- Bug fixing on SearchView and FilterMethods.
- Resolved crash bug when the sticky header is collapsed by preventing new header to be sticky. RecyclerView raised an exception not under my control.

**Demo App**
- Added an example of `FlexibleItemHolder` to use simultaneously the same _modelData_ object in multiple RecyclerView/Adapters.
- New example app supporting multiple fragments #19. Use of NavigationView and VectorDrawables.
- Adapted code for Expandable Sections with headers. Enabled demo option for `FragmentExpandableSections`.
- Demo app has more clear DatabaseService. It shows how to create item interfaces.
- Unfortunately the demo app has an _unknown_ bug on FastScroller when changing Fragment, but FastScroller is planned to be changed.

**Comment by davideas** [comment]
New **pre-release [v5.0.0-b5](https://github.com/davideas/FlexibleAdapter/releases/tag/5.0.0-b5)** many improvements and fixes. (You can download the apk from the release page).
```
dependencies {
	compile 'eu.davidea:flexible-adapter:5.0.0-b5'
}
```

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. Related Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases).

**Deprecation**
- Removed `isSelected()` and `setSelected()` from `IFlexible` interface
- Deprecated and removed `getSectionableOf()` in favor of the new `getSectionItems()`.
- Improved instance state on rotation.

**Improvements**
- New methods: `addSection()`, `addItemToSection()` (not tested yet).
- Modified `DividerItemDecoration` to show a gap between the sections #60.
- Improved `isEnabled()` on `IExpandable` items.

**Multi-level expansion**
- Added possibility to chain Expandable items into others expandable.
- Item restoration works in multi-level expansion: expanded children are forced to collapse before restoration.
- Selection coherence in multi-level expansion.

**FastScroller**
- Added a small gap between bubble and handle.
- Auto handling a change of a new LayoutManager when `FastScroller` is configured. Now FastScroller should work in every situations.
- Added SmoothScroll for `GridLayoutManager` and optimized smooth scrolling classes.

**Fixes**
- Fixed the delay on the 1st item when animating the view.
- Fixed inconsistency bug when expanding/collapsing _sticky_ header items _not_ fully visible:
- Fixed StickyHeaders with `GridLayoutManager` when all sub-items of a section fit the single row of span count #61.
- Fixed NPE when setting a null `FastScroller` instance #66.

**Demo App**
- Big code reorganization using fragments (unfinished!)
- Fixed issue #64 for demo app for API < 23.

**Comment by davideas** [comment]
New **pre-release [v5.0.0-b6](https://github.com/davideas/FlexibleAdapter/releases/tag/5.0.0-b6)** many improvements and fixes. (You can download the apk from the release page).
```
dependencies {
	compile 'eu.davidea:flexible-adapter:5.0.0-b6'
}
```

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. Related Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases).

**Deprecation**
- Experimenting `HashSet` instead of `ArrayList` for `mSelectedPositions`. As consequence `getSelectedPositions()` returns a _copied_ List from the Set (already sorted, no duplication)!

**Improvements**
- Dragging and swiping can be enabled by type #80. Items are NOT Draggable and Swipeable by default.
- Added class `UndoHelper.SimpleActionListener` to simplify the development
- `Utils.colorAccent` is now public

**Fixes**
- Fixed ArrayIndexOutOfBoundsException with StaggeredGridLayoutManager #87.
- Fixed headers showing after rotation #88.
- Fix for StaggeredGridLayoutManager in DividerItemDecoration.
- Fixed `addItemToSection` when Header is hidden: item is now added to the unique visible list.

**Demo App**
- Fix for SwipeRefreshLayout to be displayed on the top of StickyHeaders Layout #83

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. Related Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases).

**Deprecation**
- Refactored method `moveItem()` to `swapItem()`.

**Improvements**
- Alpha animation, when enabling / disabling sticky headers.
- Added automatic expand for new items `onLoadMore` (IN TESTING).
- Added new method `moveItem()` that simply moves an item without swapping them.
- Separated the initialization of expanded items at start up #86: `expandItemsAtStartUp()` must be explicitly called by the user in Activity creation phase.

**Fixes**
- Temporary solution to improve ViewHolder with sticky behaviour, (user has to pass `true` as new parameter to the FVH super.constructor). Issues #78, #79 and #92. THIS IS NOT YET THE FINAL SOLUTION.
- Fixed #94 - Item does not updated after `updateDataSet()`.

**Demo App**
- Minor changes in the demo App.
- SwipeRefresh with StickyHeader is not compatible with Dragging feature, added comments in the layout to show the difference #83.

**Comment by davideas** [comment]
Road map for final release:
- Soon I will create a new pre-release **beta7** that will published to the master branch after a small period of test including all the recent fixes from the last snapshots.
- Since library is becoming more and more stable, I will evaluate to create a pre-release **rc1** after beta7.

**Comment by briangbrown** [comment]
Looks like a sweet library. I'm going to start using this on a new project today. Thanks!

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. Related Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases).

**Deprecation**
- Deprecated method addSection() with refHeader. 

**Improvements**
- Added DrawableUtils #35
- Added new method addSection() with Comparator object.
- Improved the methods `addItemToSection()` see JavaDoc.

**Fixes**
- Method `onUpdateEmptyView()` is now called when passing null or empty new list in `updateDataSet()`. 
- Fixed #112 - NPE on originalText in `Utils.hightlightText()`.

**Demo App**
- Added new example Fragment for StaggeredLayout with multiple functionalities:
 - merge/split items at runtime (position are recalculated);
 - move item with animation taking care about the groups using the new function calculatePositionFor();
 - automatic sorting when adding new items using the new function addItemToSection();
 - example of Comparator object. - change Background at runtime, maintaining the ripple effect (no XML is used).

**Comment by davideas** [comment]
New **pre-release [v5.0.0-b7](https://github.com/davideas/FlexibleAdapter/releases/tag/5.0.0-b7)** many improvements and fixes. (You can download the apk from the release page). With this pre-release, the code is available in **master** branch as well.
```
dependencies {
	compile 'eu.davidea:flexible-adapter:5.0.0-b7'
}
```

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. No Apk of this version is done.

**Deprecation and Refactoring**
- Refactoring behaviour of `setPermanentDelete()`: default value is now true. Creating the `UndoHelper` instance, the value will be set to `false` automatically.

**Improvements**
- Added `@since` for all methods
- Added new method `getSectionItemPositions()`. It resolves #132.
- Added new method `setFilterAnimationLimit()`
- Improved performance during filtering of high numbers of items. This resolves #133 part 1: call `notifyDataSetChanged()`. Animation during filtering will be performed if the total items are below the limit of 500 (modifiable value) "_current items_" OR "_filtered items_"; Above this limit, `notifyDataSetChanged()` will be called with no animation. 
- Option to display entry animation only. See #125.
- Method in `FlexibleViewHolder.setDragHandleView()` is not final anymore!
- Upgrade to Android N (API 24), SupportLib 24.0.0, Java8.

**Fixes**
- Fixed #128 - Behavior for which `getStickySectionHeadersHolder()` should return a ViewGroup with LayoutParams coherent and initialized by the user who knows how these params has to be initialized. An exception will be raised if params are nulls.
- Fixed #126 - `onUpdateEmptyView()` is not being called in `updateDataSet() if size > 0`
- Fixed #121 - Sticky header glitch when header layout has height set to "wrap_content"
- Modified FastScroller bubble padding

**Demo App**
- BackPressed now goes back to overall View. App entry fade animation?
- StaggeredLayout is the default at startUp

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. This time the Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases). It includes all the changes of the previous SNAPSHOT plus the following.

**Deprecation and Refactoring**
- Refactored `shouldAnimate` default value from `true` to `false`. Items are not animated anymore by default, you need to enable the scrolling animations with `.setAnimationOnScrolling(true)`.
- When `setOnlyEntryAnimation()` is set true, `shouldAnimate` is automatically set true.
- Marked some methods with `//TODO: deprecation?`. Those methods are under evaluation and could disappear in the final release.

**DividerItemDecoration**
- Added new method `withOffset()`.
- Fixing NPE when no divider is provided.
- Fixed normal item offset.

**Improvements**
- Added new method `setNotifyMoveOfFilteredItems()`.
- Experimental: Added AsyncTask with synchronization animations. Fixed #133 part 2: AsyncFilter. Really lot of items can be filtered in a very short time. Adapter need to be configured accordingly, new Fragment in the demoApp helps to understand how to.
- Experimental: `updateDataSet()` is now run Asynchronously too.
- Experimental: Improved performance when `collapseAll()` with big list.
- Experimental: Improved performance on `showAllHeader()` and `hideAllHeaders()`.

**Fixes**
- Fixed #135 - The AsyncTask changes made obsolete the fix of missing parenthesis generating a NPE.
- Fixed some NPEs when null is passed to the constructor.
- Fixed scrolling animation flags.

**Demo App**
- Experimental: Added `SplashActivity`.
- Added new `FragmentAsyncFilter`, here you can customize the Adapter and the database dynamically.
- Added `DatabaseType` enumeration and created `ConfigurationDatabase`.
- Changed layout for drawer and updated the menu with check boxes.

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. The Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases). It includes all the changes of the previous SNAPSHOT plus the following.

**Deprecation and Refactoring**
none.

**Improvements**
- #144 - Adapter initialization and `updateDataSet()` with no notifyItemXXX() on headers.
- #146 - Improved calls to `setDisplayHeadersAtStartUp()` and `showAllHeaders()` to avoid duplication of headers by a non correct initialization. As consequence 2 effects are now possible when calling `setDisplayHeadersAtStartUp()` read method comment, documentation will follow.
- Adapter can be initialized with a null or empty list: `addItems()` can be safely called, but no scrolling animation on loading can be performed.
- Expand items at startup can now expand all Levels of Expandable, if expanded is initially set true. `updateDataSet()` is also compatible with the last status of sub levels.

**Fixes**
- Fixed #143 - `calculatePositionFor()` doesn't work correctly when used with `addItem()`.
- When sticky header is enabled, element header at position 0, can now be dragged: layout for sticky header is cleared out, before it was not possible.

**Demo App**
- Reviewed menu items options with check boxes.
- Added menu items persistent through some example fragments.
  - For update data set (refresh to see different result);
  - For scrolling animation (reopen the page from the drawer menu to see different result).
- More JUnit test for ItemComparator
- Fixed NPE when closing app before the execution of Runnable.

**Comment by davideas** [comment]
New **pre-release [v5.0.0-b8](https://github.com/davideas/FlexibleAdapter/releases/tag/5.0.0-b8)** many improvements and fixes. (You can download the apk from the release page).
```
dependencies {
	compile 'eu.davidea:flexible-adapter:5.0.0-b8'
}
```

**Comment by frozts91** [comment]
Hi daviddeas, i get this error while i add your library to my project.

Error:(24, 13) Failed to resolve: eu.davidea:flexible-adapter:5.0.0-b8

i use jcenter for repository

**Comment by davideas** [comment]
@frozts91, now should work as normal. The plugin didn't upload at the correct position, before.

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. No new Apk available. It includes the release **beta8**, plus the following.

**Fixes**
- Fixed #176 - Drop event, without swapItems on move (ItemTouchHelperCallback can now be set)
- Fixed #191 - AsyncFilter fatal error (Empty string will skip the search)
- Fixed #184 - Problem with endless scrolling and LOAD_MORE_RESET delay

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. The Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases). It includes the release **beta8**, plus the following.

**Improvements**
- Added methods clear() clearAllBut(viewTypesToRetain)
- Added method `addItem(item)`.
- Reviewed how the internal `Handler` can be overridden: new internal class `HandlerCallback` has been created.
- Optimized selection coherence for a parent with a selected sub item at any level.
- Reviewed javaDoc.

**Fixes**
- Fixed collapsing bug for a parent with selected sub items.
- Fixed #229 - Calling hideAllHeaders() might lead to NullPointerException (Fixed NPE and modified behaviour when hiding headers: now, it doesn't disable anymore the sticky flag)

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. The Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases). It includes the release **beta8**, all previous points, and the following (**rc1 ready**).
> **Note:** If you come from previous versions, update your code following the Wiki page [Migrations](../wiki/Migrations).

**New features**
- Resolves #236 - [Scrollable Headers and Footers](https://github.com/davideas/FlexibleAdapter/wiki/5.x-%7C-Scrollable-Headers-and-Footers): items that lay respectively at the top and at the bottom of the main items.
- Resolves #236 brand new Endless Scrolling with `pageSize` and `targetCount`. Wiki page will be adapted shortly.

**Deprecation**
- Refactored and Deprecated all marked methods in issue #200.

**Improvements**
- Resolves #227 - Elevated header items and Elevated sticky header items.
- Resolves #228 - No XML layout configuration for sticky headers.
- Resolves #201 - Added method `shouldNotifyParentOnClick()`.
- Runtime background, changing state has short enter-exit fade animation.
- Added version log at startup and Info on animation when changing LayoutManager.
- Upgrade to API 25.

**Fixes**
- Resolved #244, header item doesn't swap if it matches with the first layout position.
- Decided to adopt Solution 2 for scrolling animation; Solution 1 has been removed. Fixed a couple of scrolling animation issues, now scrolling are animated properly in all situations. Scrollable Headers will now scroll animate. Scrollable Footers cannot scroll animate, when inserted the very first time.
- Fix for `adjustSelection()` when adding new items and some items are already selected.

**Demo App**
- Removed some calls of deprecated methods.
- Modified demoApp to showcase Scrollable Expandable Headers and Footers.
- Modified demoApp to showcase the new Endless Scroll.

**Comment by davideas** [comment]
First **release candidate [v5.0.0-rc1](https://github.com/davideas/FlexibleAdapter/releases/tag/5.0.0-rc1)**: new features, improvements and fixes. (You can download the apk from the release page).

```
dependencies {
    compile 'eu.davidea:flexible-adapter:5.0.0-rc1'
}
```

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published.

**Improvements**
Added generic type param when returning FlexibleAdapter object
Resolved #269 - Setting to load more at startup when adapter list is initially empty/null.
Resolved #281 - Delegate spanSize to IFlexible item
Resolved #288 - UndoHelper.remove(...) return type reverted to Snackbar

**Fixes**
Fixed #268 - Sticky headers don't work when all headers are initially collapsed
Fixed #284 - removeAllScrollableFooters() seems to notify incorrect item range

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published.

**Improvements**
Resolved #293 - getItemViewType now returns zero when item not found.

**Fixes**
Fixed #299 - ClassCastException after click on expanded sticky header when AutoCollapse is enabled.
Fixed #296 - AutoCollapseOnExpand(true) seems not notify the header.
Fixed #295 - Items not notified on auto collapse (using normal List instead of Hash to avoid Inconsistency detection. It will be slower on big List when collapsing).
Fixed #292 - HandlerCallback mistakenly erases default Runnables posted on it's `mHandler`.

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published. **With this update you MUST review your code, see deprecations below here** (I will update the migration notes soon).

**Deprecations / New Behaviours**
- Adapter now makes a copy of the provided list at _startup_, _updateDataSet_ and _filterItems_, before applying features (from support #316).
- `onPostUpdate()` and `onPostFilter()` now require to call `super()` otherwise the emptyView won't get notified!

**Improvements**
- Optimized `updateDataSet()` and `filterItems()`: now hidden items are restored before applying the feature, so they won't get notified for removing and adding same items, but only for updating. This will improve performance avoiding useless notifications.
- Resolved #321 - Now Items can have option to be notified for a change if unfiltered.
- Resolved #293 - `getItemViewType` now returns zero when item not found.
- #34 New Customizable Fastscroller is now available, thanks to [arpinca](https://github.com/arpinca).
- Created new method `isFiltering()`.
- Increased `ANIMATE_TO_LIMIT` to `1000` items.
- Support lib to 25.3.1 (from question #331).

**Fixes**
- Fixed #328 - Filtering issue during delete search query (make sure index is in bounds).

**Notes**
- Implementing _Parcelable_ to abstract adapter items may lead to raise exception (see #324):
> java.lang.IllegalStateException: onCreateViewHolder() is not implemented. If you want FlexibleAdapter creates and binds ViewHolder for you, you must override and implement the method .createViewHolder().

That because _Parcelable_ will create a new abstract object added in the adapter list, that doesn't know anything about the last overridden methods of the real items.

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published.
> **Note:** With this update you **MUST** review your code, see deprecations below here and if you come from previous versions/snapshots, update your code following the Wiki page [Migrations](../wiki/Migrations).

> The Apk is available under the [latest release](https://github.com/davideas/FlexibleAdapter/releases). It includes the release **rc1**, all previous points, and the following:

**Deprecations / New Behaviors**
- Resolved #344 - Get rid of Method B to create and bind ViewHolders.
- Renamed `DividerItemDecoration` to `FlexibleItemDecoration`.
- Renamed `Utils` to `FlexibleUtils`.

**Improvements**
- Resolved #302 - Allow to use _custom_ LayoutManagers on the RecyclerView.
- Resolved #348 - Customizable minimum scroll amount to show the FastScroller.
- Resolved #343 - Allow to unbind ViewHolder.

**Fixes**
> Nothing useful to mention.

**Notes**
- If swipe does not properly animate: please be sure that FrontView is at lower level in the layout to support API < 21.

**Comment by davideas** [comment]
Second **release candidate [v5.0.0-rc2](https://github.com/davideas/FlexibleAdapter/releases/tag/5.0.0-rc2)**: new features, improvements and fixes. (You can download the apk from the release page).

```
dependencies {
    compile 'eu.davidea:flexible-adapter:5.0.0-rc2'
    compile 'eu.davidea:flexible-adapter-binding:1.0.0-b1'
}
```

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published.

**Deprecations / New Behaviors**
- Resolved #400 - Removal of all deprecated functions.
- `setStickyHeaderElevation` now accepts only in dpi value.

**Improvements**
- Resolved #393 - Provided last sticky header position in the callback `OnStickyHeaderChangeListener`.
- Resolved #397 - Support for fully recursive lists (expansion, collapsing, and update).
- Resolved #401 - Multilevel option to collapse or keep expanded status to sub-expandables.
- SmoothScroll layout managers now implement `IFlexibleLayoutManager`.
- `getBubbleText()` in IFlexible.
- `BubbleTextCreator` can be set in FastScroller.
- Added Log when attaching/detaching RV.

**Fixes**
- Fixed #387 - proguard issue with "SourceCode" and line numbers.
- Fixed #389 - Removed advancement of position beyond subItems such that now all subItem positions will be checked for nested expansion during update.
- Fixed #392 - FlexibleItemDecoration withOffset not working correctly for section headers (use of spanSize).
- Fixed #394 - Improved expandable methods.
- Fixed #394 - Add subItem doesn't account for others expanded subItems.
- Fixed #395 - Use of MarginLayoutParams instead of LayoutParams when init StickyHeadersHolder layout.
- Fixed #404 - Expandable Sticky Header becomes invalid after updateDataSet.

**Comment by davideas** [comment]
## State of work (next Snapshot release)
The following changes have not been release yet.

### Temporary Revised `UndoHelper`
> ⚠️ **Warning:** This is a temporary state of work and _might_ change in the near future, depends by the work in progress to support the new swipe capability.
- Resolves #438 - Deprecated `OnActionListener` in UndoHelper also `onPreAction()` and `onPostAction()`.<br>New comprehensive methods name: Renamed the 2 callbacks methods in `OnUndoListener` and renamed also the method `remove` series:
``` java
.withAction(int action, OnActionListener listener) >>> .withAction(@Action int action)
.remove(...) >>> .start(...) // Still returns Snackbar object

OnActionListener >>> !!deprecated!! // Please read related javaDoc

OnUndoListener.onDeleteConfirmed(int action) >>> onActionConfirmed(@Action int action, int event)
OnUndoListener.onUndoConfirmed(int action) >>> onActionCanceled(@Action int action)
```
- Resolves #434 - Consecutive Undo requests option & improved the general stability. New method to commit one by one the swiped items and undo only the latest action. Also, it will clear the internal memory references at the end of the dismissal event!
``` java
.withConsecutive(true/false) // Default value is false (accumulate items to undo)
```
- Added the type of **event** with the following cases:
``` java
onActionConfirmed(@Action int action, int event);
- DISMISS_EVENT_SWIPE = 0;       // Snackbar was dismissed via a swipe
- DISMISS_EVENT_TIMEOUT = 2;     // Snackbar was dismissed via a timeout.
- DISMISS_EVENT_MANUAL = 3;      // Snackbar was dismissed via a call to dismiss().
- DISMISS_EVENT_CONSECUTIVE = 4; // Snackbar was dismissed from a new Snackbar being shown.
```
- While the `onActionCanceled` is always called because of event type "action" only (user pressed on snackbar action button) and therefore not provided.
``` java
onActionCanceled(@Action int action);
- DISMISS_EVENT_ACTION = 1;      // Snackbar was dismissed via an action click.
```

### Improvements
- Resolved #431 - CustomTags: Different TAGs for each Adapter instance.
- Resolved #428 - Option to disable swipe/drag capabilities when ActionMode is activated.
- Resolved #427 - Dynamically add and remove listeners.
- Scripts with GrabVer 0.3.0.

### Fixes
- Fixed #409 - `setEndlessScrollListener` crashes when logging is enabled.

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published.

### Improvements
- Upgrade to Android Studio 3.0.0 and Support Lib 27.0.0.

### Fixes
- Fixed #475 - custom and default divider gone in rc3.
- Fixed #477 - ExpandableViewHolder#onClick should always call super().
- Fixed #481 - highlightText and highlightWords text looses the upper case during the scan.
- Fixed #484 - bindViewHolder method is causing a problem with GreenDao.

**Comment by amankumarjain** [comment]
Using:
repositories {
        jcenter()
}

Error:
Failed to resolve: eu.davidea:flexible-adapter:5.0.0-rc3

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published.
New **beta2 UI** and **beta2 LiveData** extenstions.

### Improvements
- Resolved #384 - Ability to recognize child view click.
- Resolved #507 - New filter results callback: support for observing the current state of filtering.
- Resolved #523 - Swipe delete, `UnsupportedOperationException` for immutable list on sort.
- Added new method `setHeadersShown(boolean)` to manually change the flag to indicate that headers are already inserted in the main list by the user.
- Removed deprecated functions in AnimatorAdapter.
- Improved nullity of FastScroller instance.

### Fixes
Fixed #512 - NPE in `initStickyHeadersHolder` on orientation change.
Fixed #521 - Fixed Sticky Headers swapping with Staggered Layout.
Fixed #516 - Better section gap when expand/collapse the last item of section.
Fixed #514 - Removed @NonNull annotation for parameter in `addListener` and `removeListener` methods.

### UI extension
- Added EmptyViewHelper.

### LiveData extension
- Added Comparator to sort items when using FlexibleItemProvider with LiveData.

**Comment by davideas** [comment]
New **5.0.0-SNAPSHOT** has been published.
New **beta2 UI** and **beta2 LiveData** extenstions.

### Improvements
- Resolved #384 - Ability to recognize child view click.
- Resolved #507 - New filter results callback: support for observing the current state of filtering.
- Resolved #523 - Swipe delete, `UnsupportedOperationException` for immutable list on sort.
- Resolved #540 - Multi filter.
- Resolved #543 - `DiffUtil` is back.
- Added new method `setHeadersShown(boolean)` to manually change the flag to indicate that headers are already inserted in the main list by the user.
- Removed deprecated functions in AnimatorAdapter.
- Improved nullity of FastScroller instance.

### Fixes
Fixed #512 - NPE in `initStickyHeadersHolder` on orientation change.
Fixed #521 - Fixed Sticky Headers swapping with Staggered Layout.
Fixed #516 - Better section gap when expand/collapse the last item of section.
Fixed #514 - Removed @NonNull annotation for parameter in `addListener` and `removeListener` methods.

### UI extension
- Added EmptyViewHelper #439.

### LiveData extension
- Added Comparator to sort items when using FlexibleItemProvider with LiveData.
